internal/future/serving/server.go

```diff
@@ -90,3 +129,78 @@ func (p *Params) AddHandleFunc(handlerFunc GrpcHandler, grpcProxyHandler GrpcPro
 	}
 	return p
 }
+
+// invalidate closes all the TCP listeners that would otherwise leak if initialization fails.
+func (p *Params) invalidate() {
+	if err := p.grpcListener.Close(); err != nil {
+		serverLogger.Errorf("error closing grpc handler, %s", err)
+	}
+	if err := p.grpcProxyListener.Close(); err != nil {
+		serverLogger.Errorf("error closing grpc-proxy handler, %s", err)
+	}
+}
+
+// Server hosts a gRPC and HTTP server.
+// All HTTP traffic is served from a common http.ServeMux.
+type Server struct {
+	serverWithProxy grpcServerWithProxy
+}
+
+// grpcServerWithProxy this will go away when insecure.go and tls.go are merged into the same server.
+type grpcServerWithProxy interface {
+	start(*Params) (func(), error)
+	stop()
+}
+
+// Start the gRPC+HTTP(s) REST server.
+func (s *Server) Start(p *Params) (func(), error) {
+	if p.usingTLS() {
+		s.serverWithProxy = newTLSServer(p.grpcListener, p.grpcProxyListener)
+	} else {
+		s.serverWithProxy = newInsecureServer(p.grpcListener, p.grpcProxyListener)
+	}
+	return s.serverWithProxy.start(p)
+}
+
+// Stop the gRPC+HTTP(s) REST server.
+func (s *Server) Stop() {
+	s.serverWithProxy.stop()
+}
+
+// New creates an instance of an Open Match server.
+func New() *Server {
+	return &Server{}
+}
+
+func startServingIndefinitely(params *Params) (func(), func(), error) {
```

Comment:
is this helper for testing purposes so you can get the helper to kill from tests? the name was confusing - as it actually does not start serving but returns a method that should be called to start serving? can you add a comment explaining that - or more intuitive name?

Reply:
The method starts the serving process it returns 2 functors.
`serveUntilKilledFunc()` which will sleep wait until Ctrl+C is pressed.
`stopServingFunc()` which will simulate Ctrl+C being pressed which will terminate `serveUntilKilledFunc()`.

By the time `startServingIndefinitely()` has finish the gRPC and proxy can serve traffic because `waitForServing()` inside of it was called.

Reply:
Thanks! Can you add that in a comment please?